### PR TITLE
chore: LockTable tracks fingerprints of keys

### PR DESCRIPTION
**Pull request by romange**
It's a first step that will probably simplify dependencies in many places where we need to keep key strings for that. A second step will be to reduce the CPU load of multi-key operations like MSET and precompute Fingerprints once.

### CODE REVIEW

File: src/server/table.h [hunk 1]
```diff
@@ -99,26 +101,13 @@ class LockTable {
...
     }
-
-    // If the key is backed by a string_view, replace it with a string with the same value
-    void MakeOwned() const;
-
-    mutable std::variant<std::string_view, std::string> val_;
   };
-
-  absl::flat_hash_map<Key, IntentLock> locks_;
+  absl::flat_hash_map<uint64_t, IntentLock, Hasher> locks_;
```
Comment (chakaz) [inline src/server/table.h:110]: Why do we need a custom hasher instead of just `absl::flat_hash_map<uint64_t, IntentLock>`?

File: src/server/table.h [hunk 2]
```diff
@@ -90,6 +90,8 @@ class LockTable {
   bool Acquire(std::string_view key, IntentLock::Mode mode);
   void Release(std::string_view key, IntentLock::Mode mode);
 
+  static uint64_t Fingerprint(std::string_view key);
```
Comment (chakaz) [inline src/server/table.h:93]: Make private?

File: src/server/transaction.h [hunk 3]
```diff
@@ -500,6 +498,7 @@ class Transaction {
   // Multi transactions unlock asynchronously, so they need to keep a copy of all they keys.
   // "Launder" keys by filtering uniques and replacing pointers with same lifetime as transaction.
   void LaunderKeyStorage(CmdArgVec* keys);
+  bool CheckLock(EngineShard* shard, IntentLock::Mode mode, const KeyLockArgs& lock_args);
```
Comment (dranikpg) [inline src/server/transaction.h:501]: 1. Please add a comment
2. Why did you move it here? Why do Release and Acquire work with KeyLockArgs and this one is separate? More exceptions - > more diffciult code
3. Confusing signature - this function is a member, but doesn't access anything

Comment (dranikpg) [reply src/server/transaction.h:501]: Moreover, KeyLockArgs if free to construct for, why bother with separation

File: src/server/table.h (hunk 2 above)
Comment (dranikpg) [reply src/server/table.h:93]: Or even more it to cc

**Review by dranikpg — COMMENTED** [review]
Please share benchmark results. I like that we remove code, but I doubt it's faster, we just moved the hashing part to our side

File: src/server/table.h (hunk 1 above)
Comment (romange) [reply src/server/table.h:110]: line 104 explains. If it's not clear I will expand there.

File: src/server/transaction.h (hunk 3 above)
Comment (romange) [reply src/server/transaction.h:501]: I removed the function in DbSlice. I removed the code,  not added it.

**Comment by romange** [comment]
> Please share benchmark results. I like that we remove code, but I doubt it's faster, we just moved the hashing part to our side

Exactly right. Because it's a half-work. Doing the second half :)

File: src/server/transaction.h (hunk 3 above)
Comment (dranikpg) [reply src/server/transaction.h:501]: You _moved_ it here 🙂 More specifically, to be a member. I'm asking why

**Comment by romange** [comment]
Ok,  I implemented Release/Acquire with fingerprints. the performance difference is negligible for MSET.
@dranikpg I am fine to sack this PR but I think it will simplify things for multi transactions as well.
See the second half here: https://github.com/dragonflydb/dragonfly/compare/Pr6?expand=1

**Review by dranikpg — COMMENTED** [review]
Okay, so given it doesn't have much of performance advantages

1. I'm not opposed to the idea if it simplifies code. _However_, in many cases we need to now keep track of both string and fingerprints, not sure if it's a simplification on average over the long run.
2. I still don't understand why you insist on making CheckLock a member function on transaction. It never belonged there, especially _as a member_. We spent so much effort previously to remove all excessive code from it 😿

**Review by dranikpg — DISMISSED** [review]
You can move (2) to anonymous namespace. I'm being annoyingly picky because of https://en.wikipedia.org/wiki/Broken_windows_theory. First glass shattering 😆 

(1) Is up to you. I don't have strong opinions as long as it's maintainable

File: src/server/table.h (hunk 2 above)
Comment (romange) [reply src/server/table.h:92]: Done

File: src/server/transaction.h (hunk 3 above)
Comment (romange) [reply src/server/transaction.h]: removed

**Comment by romange** [comment]
I decided to proceed with this PR. I think this PR will help me to implement more advanced heuristics of VLL paper, where we can identify OOO Transactions in the TX queue even before they reach the TX-queue head.  Not something we must do urgently but I think it does simplify things with everything related to key locking.

File: src/server/transaction.cc [hunk 4]
```diff
@@ -46,12 +46,12 @@ void AnalyzeTxQueue(const EngineShard* shard, const TxQueue* txq) {
     if (now >= last_log_time + 10) {
       last_log_time = now;
       EngineShard::TxQueueInfo info = shard->AnalyzeTxQueue();
-      string msg =
-          StrCat("TxQueue is too long. Tx count:", info.tx_total, ", armed:", info.tx_armed,
-                 ", runnable:", info.tx_runnable, ", total locks: ", info.total_locks,
-                 ", contended locks: ", info.contended_locks, "\n");
+      string msg = StrCat("TxQueue is too long. Tx count:", info.tx_total,
+                          ", armed:", info.tx_armed, ", runnable:", info.tx_runnable,
+                          ", global:", info.tx_global, ", total locks: ", info.total_locks,
```
Comment (romange) [inline src/server/transaction.cc:51]: just added global stats

File: src/server/table.h [hunk 5]
```diff
@@ -92,17 +89,24 @@ class LockTable {
   bool Acquire(std::string_view key, IntentLock::Mode mode);
   void Release(std::string_view key, IntentLock::Mode mode);
```
Comment (chakaz) [inline src/server/table.h:90]: Here's a proposal: have overloads for both of these functions, the existing one (accepting a `string_view`), and another one accepting a fingerprint.
Then have the `string_view` version do the hash and call the overloaded one. This will save you the trouble of adding them in the future PR you plan to add.

File: src/server/table.h [hunk 6]
```diff
@@ -92,17 +89,24 @@ class LockTable {
...
 
+  static uint64_t Fingerprint(std::string_view key);
```
Comment (chakaz) [inline src/server/table.h:92]: Consider some typedef for the fingerprint type, like `using THash = uint64_t`

**Review by chakaz — APPROVED** [review]

File: src/server/table.h (hunk 5 above)
Comment (romange) [reply src/server/table.h:90]: I already wrote this in another branch. I prefer adding them in the next PR but I did like you suggested.

File: src/server/table.h (hunk 6 above)
Comment (dranikpg) [reply src/server/table.h]: It won't grow it won't shrink, keeps less types. Std uses size_t as well after all

Comment (chakaz) [reply src/server/table.h]: giving a name to this "type" makes the code easier to read (like that it's not an index but a hash). once C++ will get distinct typedef it'll be even better, because then the compiler can guard against such errors

**Review by chakaz — APPROVED** [review]

**Review by chakaz — APPROVED** [review]
